File: src/cviz/Timeline.java
package cviz;

import cviz.state.State;
import cviz.timeline.Trigger;
import cviz.timeline.TriggerType;
import cviz.timeline.commands.ClearCommand;
import cviz.timeline.commands.ICommand;
import se.svt.caspar.amcp.AmcpChannel;
import se.svt.caspar.amcp.AmcpLayer;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

public class Timeline implements ITimeline, Runnable {
    private final String timelineId;
    private final AmcpChannel channel;
    private final LinkedList<Trigger> remainingTriggers;
    private final CopyOnWriteArrayList<Trigger> activeTriggers = new CopyOnWriteArrayList<>();
    private final ConcurrentHashMap<Integer, LayerState> currentLayerState = new ConcurrentHashMap<>();
    private final Set<Integer> usedLayers = new HashSet<>();
    private final HashMap<Integer, AmcpLayer> layerCache = new HashMap<>();

    private final State state;

    private HashMap<String, String> parameterValues;
    private boolean running = false;
    private boolean killNow = false;

    public Timeline(String timelineId, AmcpChannel channel, State state, LinkedList<Trigger> triggers) {
        this.timelineId = timelineId;
        this.channel = channel;
        this.state = state;
        this.remainingTriggers = triggers;

        state.setState(TimelineState.READY);
    }

    public int getChannelNumber(){
        return channel.channelId();
    }

    @Override
    public AmcpLayer getLayer(int layerId) {
        return layerCache.computeIfAbsent(layerId, k -> new AmcpLayer(channel, layerId));
    }

    public void kill() {
        System.out.println("Timeline " + timelineId + " received kill");
        killNow = true;
        running = false;
    }

    @Override
    public void setLayerState(int layerId, LayerState state) {
        currentLayerState.put(layerId, state);
    }

    @Override
    public LayerState getLayerState(int layerId) {
        return currentLayerState.get(layerId);
    }

    public State getState() {
        return state;
    }

    @Override
    public CopyOnWriteArrayList<Trigger> getActiveTriggers() {
        return activeTriggers;
    }

    public boolean isRunning() {
        return running;
    }

    private Optional<Trigger> getCueTrigger() {
        Optional<Trigger> next =  activeTriggers.stream().filter(t -> !t.isLoop()).findFirst();
        if (!next.isPresent())
            return next;

        if(next.get().getType() == TriggerType.CUE)
            return next;

        return Optional.empty();
    }

    private boolean areRequiredParametersDefined() {
        HashSet<String> fields = getParameterNames(remainingTriggers);
        for (String fieldName : fields) {
            if (fieldName.indexOf("@") == 0 && !parameterValues.containsKey(fieldName.substring(1))) {
                state.setState(TimelineState.ERROR, "Missing required parameter: " + fieldName);
                return false;
            }
        }
        return true;
    }

    public static HashSet<String> getParameterNames(LinkedList<Trigger> triggers){
        HashSet<String> fields = new HashSet<>();

        for (Trigger t : triggers) {
            for (ICommand c : t.getCommands()) {
                fields.addAll(Arrays.asList(c.getParameters()));
            }
        }

        return fields;
    }

    @Override
    public void run() {
        if (running) return;
        running = true;

        // check all required parameters are defined
        if (!areRequiredParametersDefined()) {
            running = false;
            return;
        }

        // run any setup triggers
        long setupTriggerCount = remainingTriggers.stream().filter(t -> t.getType() == TriggerType.SETUP).count();
        if (setupTriggerCount > 1){
            System.out.println("Timeline can only have one setup trigger");
            running = false;
            return;
        }

        // collect the list of layers being altered
        for (Trigger t : remainingTriggers) {
            for (ICommand c : t.getCommands()) {
                usedLayers.add(c.getLayerId());
            }
        }

        System.out.println("Template spans " + usedLayers.size() + " layers");

        Trigger setupTrigger = null;
        if (remainingTriggers.peekFirst().getType() == TriggerType.SETUP)
            setupTrigger = remainingTriggers.pop();

        state.setState(TimelineState.RUN);

        System.out.println("Starting timeline " + timelineId);

        // set some triggers as active
        promoteTriggersToActive();

        // Run the setup
        if (setupTrigger != null)
            executeTrigger(setupTrigger);

        while (running) {
            synchronized (this) {
                if (remainingTriggers.isEmpty() && activeTriggers.isEmpty())
                    break;
            }

            Optional<Trigger> cueTrigger = getCueTrigger();
            if (cueTrigger.isPresent())
                state.setState(TimelineState.CUE, cueTrigger.get().getName());
            else
                state.setState(TimelineState.RUN);

            // wait until the timeline has been finished
            try {
                Thread.sleep(10);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        // if kill command has been sent, then wipe everything
        if (killNow) {
            state.setState(TimelineState.ERROR, "Killed");
            remainingTriggers.clear();
            activeTriggers.clear();
        }

        //ensure everything has been reset
        clearAllUsedLayers();

        System.out.println("Finished running timeline");
        running = false;
        state.setState(TimelineState.CLEAR);
    }

    private void clearAllUsedLayers() {
        for (Integer l : usedLayers) {
            ICommand c = new ClearCommand(l);
            c.execute(this);
            System.out.println("Clearing layer " + l);
        }
    }

    private boolean promoteTriggersToActive() {
        int moved = 0;

        while (!remainingTriggers.isEmpty()) {
            Trigger t = remainingTriggers.pop();
            moved++;

            activeTriggers.add(t);

            // we only want to add up until a manual trigger
            if (t.getType() == TriggerType.CUE)
                break;
        }

        return moved > 0;
    }

    public synchronized void triggerCue() {
        if (!running) {
            System.err.println("Received cue when not running");
            return;
        }
        System.out.println("Received a cue");
        // TODO - maybe this should be buffered, otherwise there could be some timing issues

        state.setState(TimelineState.RUN);

        // find trigger to cue
        Optional<Trigger> waiting = getCueTrigger();
        if (!waiting.isPresent()) {
            System.err.println("Received a cue without a trigger to fire");
            return;
        }

        // run the trigger
        executeTrigger(waiting.get());
        activeTriggers.remove(waiting.get());

        if (!promoteTriggersToActive()) {
            System.out.println("Reached end of timeline");
        }
    }

    synchronized void triggerOnVideoFrame(int layer, long frame, long totalFrames) {
        if (!running) return;

        for (Trigger t: activeTriggers) {
            if (t.getType() == TriggerType.SETUP)
                continue;

            if (t.getLayerId() != layer)
                continue;

            // determine the frame we are aiming for
            long targetFrame = totalFrames;
            if (t.getType() != TriggerType.END) {
                targetFrame = t.getTargetFrame();
            }

            LayerState state = currentLayerState.get(layer);
            if (state == null) {
                System.err.println("Tried to get state and failed");
            }
            // TODO - this check needs to ensure that an appropriate amount of time has passed
            // NOTE: this also gets hit if the source video is a different framerate to the channel
            else if (state.getLastFrame() == frame && targetFrame > frame) {
                // the video didn't play to the end for some reason, move on
                System.err.println("Loop didn't reach the end, check your video!");

                executeTrigger(t);
            } else if (t.hasWaited()) {
                // do it
                executeTrigger(t);
            } else if (frame >= targetFrame) {
                t.setWaited();
            }
        }

        if (currentLayerState.containsKey(layer)) {
            currentLayerState.get(layer).setLastFrame(frame);
        }
    }

    private void executeTrigger(Trigger trigger) {
        // TODO - may want to look into using a thread to do the sending/commands, as they block until they get a response
        // may cause issues with integrity of remainingTriggers lists though
        trigger.getCommands().forEach(c -> c.execute(this));
        activeTriggers.remove(trigger);
    }

    public String getParameterValue(String name, boolean escape) {
        if (name.indexOf("@") == 0) {
            String param = parameterValues.get(name.substring(1));
            if (param == null)
                return name;

            if (escape){
                param = param.replace("\\n", "\\\\n"); // " => \"
                param = param.replace("\\\"", "\\\\\\\\\""); // \" => \\\\"
                param = param.replace("\"", "\\\""); // " => \"
            }

            return param;
        }

        if (name.indexOf("\"@") == 0 && name.lastIndexOf("\"") == name.length()-1)
            return "\"" + getParameterValue(name.substring(1, name.length()-1), escape) + "\"";

        return name;
    }

    void setParameterValues(HashMap<String, String> parameterValues) {
        if (this.parameterValues != null)
            return;

        this.parameterValues = parameterValues != null ? parameterValues : new HashMap<>();
    }
}

File: CViz/Timeline/Timeline.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Text;
using System.Threading;
using CViz.Timeline.Command;
using CViz.Timeline.Triggers;
using CViz.Util;
using log4net;
using LibAtem.Net;
using StilSoft.CasparCG.AmcpClient;

namespace CViz.Timeline
{
    class Timeline : ITimeline
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(Timeline));

        private readonly string _timelineId;
        private readonly TimelineSpec _spec;
        private readonly ConcurrentDictionary<int, LayerState> _currentLayerState;
        private readonly HashSet<int> _usedLayers;

        private TimelineTriggerSet _previousTriggers;
        private TimelineTriggerSet _triggers;
        private readonly object _triggersLock;
        
        private int _portId;
        private long _portFirstFrame;

        public Timeline(string timelineId, AmcpConnection casparClient, AtemClient atemClient, int channelId, TimelineState state, TimelineSpec spec)
        {
            _timelineId = timelineId;
            CasparClient = casparClient;
            AtemClient = atemClient;
            ChannelNumber = channelId;
            State = state;
            _spec = spec;

            _triggersLock = new object();
            _triggers = new TimelineTriggerSet(spec.Triggers.ToList(), new List<ITrigger>(), null);

            _currentLayerState = new ConcurrentDictionary<int, LayerState>();
            _usedLayers = new HashSet<int>();

            state.SetState(TimelineState.StateType.Ready);
        }
        
        public int ChannelNumber { get; }
        public TimelineState State { get; }

        public AmcpConnection CasparClient { get; }
        public AtemClient AtemClient { get; }

        public bool IsRunning { get; private set; }
        private bool KillNow { get; set; }

        public void Kill()
        {
            Log.InfoFormat("Killing timeline: {0}", _timelineId);
            KillNow = true;
            IsRunning = false;
        }

        public void SetLayerState(int layerId, LayerState state) => _currentLayerState[layerId] = state;
        public LayerState GetLayerState(int layerId) => _currentLayerState.TryGetValue(layerId, out LayerState val) ? val : null;
        
        private CueTrigger GetCueTrigger()
        {
            lock (_triggersLock)
            {
                ITrigger next = _triggers.Active.FirstOrDefault(t => !(t is LoopTrigger));
                return next as CueTrigger;
            }
        }

        private bool AreRequiredParametersDefined(TimelineTriggerSet triggerSet)
        {
            foreach (string fieldName in triggerSet.Remaining.SelectMany(t => t.Commands).SelectMany(c => c.Parameters).Distinct())
            {
                if (fieldName.IndexOf("@", StringComparison.InvariantCulture) == 0 && !triggerSet.ParameterValues.ContainsKey(fieldName.Substring(1)))
                {
                    State.SetState(TimelineState.StateType.Error, "Missing required parameter: " + fieldName);
                    return false;
                }
            }
            return true;
        }
        
        public static HashSet<string> GetParameterNames(IEnumerable<ITrigger> triggers)
        {
            return new HashSet<string>(triggers.SelectMany(t => t.Commands).SelectMany(c => c.Parameters));
        }
        
        public void Run()
        {
            if (IsRunning) return;
            IsRunning = true;

            // check all required parameters are defined
            if (!AreRequiredParametersDefined(_triggers))
            {
                IsRunning = false;
                return;
            }

            // collect the list of layers being altered
            _usedLayers.AddRange(_triggers.Remaining.SelectMany(t => t.Commands).Select(c => c.LayerId).Where(l => l > 0));

            Log.InfoFormat("Template spans {0} layers", _usedLayers.Count);

            ITrigger setupTrigger = _triggers.Remaining.OfType<SetupTrigger>().FirstOrDefault();
            if (setupTrigger != null)
                _triggers.Remaining.Remove(setupTrigger);

            State.SetState(TimelineState.StateType.Run);

            Log.InfoFormat("Starting timeline: {0}", _timelineId);

            // set some triggers as active
            PromoteTriggersToActive();

            // Run the setup
            if (setupTrigger != null)
                ExecuteTrigger(setupTrigger);

            while (IsRunning)
            {
                lock(_triggersLock) {
                    if (!_triggers.Remaining.Any() && _triggers.Active.All(t => (t is LoopTrigger)))
                    {
                        if (_previousTriggers != null)
                            SwapBackToMainTimeline();
                        else
                            break;
                    }
                }

                CueTrigger cueTrigger = GetCueTrigger();
                if (cueTrigger != null)
                    State.SetState(
                        cueTrigger is RunChildOrCueTrigger
                            ? TimelineState.StateType.CueOrChild
                            : TimelineState.StateType.Cue, cueTrigger.Name);
                else
                    State.SetState(TimelineState.StateType.Run);

                // wait until the timeline has been finished
                try
                {
                    Thread.Sleep(10);
                }
                catch (Exception)
                {
                }
            }

            // if kill command has been sent, then wipe everything
            if (KillNow)
            {
                State.SetState(TimelineState.StateType.Error, "Killed");
                _triggers.Remaining.Clear();
                _triggers.Active.Clear();
            }

            //ensure everything has been reset
            ClearAllUsedLayers();

            Log.InfoFormat("Finished timeline: {0}", _timelineId);
            IsRunning = false;
            State.SetState(TimelineState.StateType.Clear);
        }

        private void ClearAllUsedLayers()
        {
            List<int> layers = _usedLayers.ToList();
            
            foreach (int l in layers)
            {
                CommandBase c = new ClearCommand(l);
                c.Execute(this);
                Log.InfoFormat("Clearing layer: {0}", l);
            }
        }

        private bool PromoteTriggersToActive()
        {
            lock (_triggersLock)
            {
                int moved = 0;

                while (_triggers.Remaining.Any())
                {
                    ITrigger t = _triggers.Remaining.FirstOrDefault();
                    if (t == null)
                        break;

                    moved++;

                    _triggers.Remaining.RemoveAt(0);

                    _triggers.Active.Add(t);
                    
                    // we only want to add up until a manual trigger
                    if (t is CueTrigger || t is DelayTrigger)
                        break;
                }

                if (moved == 0 && _previousTriggers != null)
                {
                    SwapBackToMainTimeline();
                    return true;
                }

                return moved > 0;
            }
        }

        private void SwapBackToMainTimeline()
        {
            lock (_triggersLock)
            {
                _previousTriggers.Active.AddRange(_triggers.Active.Where(t => t is LoopTrigger));
                _triggers = _previousTriggers;
                _previousTriggers = null;
            }
        }

        public void TriggerChild(string name, Dictionary<string, string> parameters)
        {
            lock (_triggersLock)
            {
                if (!IsRunning)
                {
                    Log.InfoFormat("Recevied runchild when not running");
                    return;
                }
                Log.InfoFormat("Recevied runchild for: {0}", _timelineId);

                State.SetState(TimelineState.StateType.Run);

                // find trigger to cue
                RunChildOrCueTrigger waiting = GetCueTrigger() as RunChildOrCueTrigger;
                if (waiting == null)
                {
                    Log.InfoFormat("Recevied runchild with no trigger to fire: {0}", _timelineId);
                    return;
                }

                if (!_spec.ChildTimelines.TryGetValue(waiting.TimelineName, out ImmutableArray<ITrigger> newTriggers))
                {
                    Log.InfoFormat("Timeline is corrupt and is missing child timeline data: {0}", _timelineId);
                    return;
                }

                var newTriggerSet = new TimelineTriggerSet(newTriggers.Select(t => t.Clone()).ToList(),
                    _triggers.Active.Where(t => t is LoopTrigger).ToList(), parameters.ToImmutableDictionary());

                // check all required parameters are defined
                if (!AreRequiredParametersDefined(newTriggerSet)) // TODO fix this usage! _parameters havent been upated yet
                {
                    Log.WarnFormat("Not all parameters to child timeline are defined. Ignoring run");
                    return;
                }

                // Promote triggers
                _previousTriggers = _triggers;
                _triggers = newTriggerSet;
                State.InstanceName = name;

                _previousTriggers.Active.RemoveAll(t => t is LoopTrigger);
                
                // collect the list of layers being altered
                _usedLayers.AddRange(_triggers.Remaining.SelectMany(t => t.Commands).Select(c => c.LayerId).Where(l => l > 0));

                ITrigger setupTrigger = _triggers.Remaining.OfType<SetupTrigger>().FirstOrDefault();
                if (setupTrigger != null)
                    _triggers.Remaining.Remove(setupTrigger);

                State.SetState(TimelineState.StateType.Run);

                Log.InfoFormat("Starting child timeline: {0}", waiting.TimelineName);

                // set some triggers as active
                PromoteTriggersToActive();

                // Run the setup
                if (setupTrigger != null)
                    ExecuteTrigger(setupTrigger);
            }
        }

        public void TriggerCue()
        {
            lock (_triggersLock)
            {
                if (!IsRunning)
                {
                    Log.InfoFormat("Recevied cue when not running");
                    return;
                }
                Log.InfoFormat("Recevied cue for: {0}", _timelineId);
                // TODO - maybe this should be buffered, otherwise there could be some timing issues

                State.SetState(TimelineState.StateType.Run);

                // find trigger to cue
                CueTrigger waiting = GetCueTrigger();
                if (waiting == null)
                {
                    Log.InfoFormat("Recevied cue with no trigger to fire: {0}", _timelineId);
                    return;
                }

                // run the trigger
                ExecuteTrigger(waiting);

                if (!PromoteTriggersToActive())
                {
                    Log.InfoFormat("Reached end of timeline: {0}", _timelineId);
                }
            }
        }

        internal void TriggerOnLayerFrame(int layer, LayerType type, long frame, long totalFrames)
        {
            lock (_triggersLock)
            {
                if (!IsRunning) return;

                // TODO - this could be rewritten to always keep currentLayerState up to date, as some triggers may need the info. but changing the timeline can fix that
                foreach (FrameTrigger t in _triggers.Active.OfType<FrameTrigger>().ToList())
                {
                    if (t.Layer != layer)
                        continue;

                    // determine the frame we are aiming for
                    long targetFrame = totalFrames;
                    if (t.TargetFrame != -1)
                        targetFrame = t.TargetFrame;

                    if (!_currentLayerState.TryGetValue(layer, out LayerState state) && type == LayerType.Scene)
                    {
                        state = new LayerState(type, "");
                        _currentLayerState[layer] = state;
                    }

                    if (state == null)
                    {
                        Log.InfoFormat("Failed to get layer state: {0}-{1}", _timelineId, layer);
                    }
                    else if (state.Type != type)
                    {
                        Log.WarnFormat("Mismatched layer typee: {0}-{1}", _timelineId, layer);
                    }
                    // TODO - this check needs to ensure that an appropriate amount of time has passed
                    // NOTE: this also gets hit if the source video is a different framerate to the channel
                    else if (state.LastFrame == frame && targetFrame > frame)
                    {
                        // the video didn't play to the end for some reason, move on
                        if (state.Type != LayerType.Scene || targetFrame - 1 != frame)
                            Log.InfoFormat("Loop didn't reach the end, check your video! {0}-{1}", _timelineId, layer);

                        ExecuteTrigger(t);
                    }
                    else if (t.Waited)
                    {
                        // do it
                        ExecuteTrigger(t);

                    }
                    else if (frame >= targetFrame)
                    {
                        t.SetWaited();
                    }
                }

                if (_currentLayerState.TryGetValue(layer, out LayerState st))
                    st.LastFrame = frame;
            }
        }
        
        internal void TriggerOnScenePaused(int layer, string stopName)
        {
            lock (_triggersLock)
            {
                if (!IsRunning) return;

                foreach (SceneStopTrigger t in _triggers.Active.OfType<SceneStopTrigger>().ToList())
                {
                    if (t.Layer != layer)
                        continue;

                    if (t.StopName != stopName)
                        continue;

                    ExecuteTrigger(t);
                }
            }
        }

        internal void TriggerOnChannelFrame(int port, long frame)
        {
            // Use the first port that is found
            if (_portId == 0)
                _portId = port;
            if (_portId != port)
                return;

            if (_portFirstFrame == 0)
                _portFirstFrame = frame;

            if (frame <= _portFirstFrame)
                return;

            lock (_triggersLock)
            {
                if (!IsRunning)
                    return;

                // find trigger to cue
                DelayTrigger next = _triggers.Active.FirstOrDefault(t => !(t is LoopTrigger)) as DelayTrigger;
                if (next == null)
                    return;

                // If the counter hasnt started, then set it as starting on the previous frame
                if (!next.DelayStarted)
                    next.StartDelay(frame - 1);

                // Check if hit the target yet
                if (next.DelayEndAt > frame)
                    return;

                ExecuteTrigger(next);

                if (!PromoteTriggersToActive())
                {
                    Log.InfoFormat("Reached end of timeline: {0}", _timelineId);
                }
            }
        }

        private void ExecuteTrigger(ITrigger trigger)
        {
            lock (_triggersLock)
            {
                // TODO - may want to look into using a thread to do the sending/commands, as they block until they get a response
                // may cause issues with integrity of remainingTriggers lists though
                trigger.Commands.ForEach(c => c.Execute(this));
                _triggers.Active.Remove(trigger);
            }
        }
        
        public string GetParameterValue(string name, bool escape)
        {
            if (name.IndexOf("@", StringComparison.InvariantCulture) == 0)
            {
                if (!_triggers.ParameterValues.TryGetValue(name.Substring(1), out string param))
                    return name;

                if (escape)
                    param = param.AddSlashes();

                return param;
            }

            if (name.IndexOf("\"@", StringComparison.InvariantCulture) == 0 && name.LastIndexOf("\"", StringComparison.InvariantCulture) == name.Length - 1)
                return "\"" + GetParameterValue(name.Substring(1, name.Length - 1), escape) + "\"";

            return name;
        }

        internal void SetParameterValues(ImmutableDictionary<string, string> parameterValues)
        {
            _triggers.SetParameters(parameterValues);
        }


        public void AddTrigger(ITrigger t)
        {
            lock (_triggersLock)
                _triggers.Active.Add(t);
        }

        public void RemoveAllTriggers(Predicate<ITrigger> predicate)
        {
            lock (_triggersLock)
                _triggers.Active.RemoveAll(predicate);
        }
        
    }
}